### ISSUE TITLE
[Feature] Support activation CPU offload with FSDP2/AutoModel

### ISSUE THREAD

**Issue by xinyu-dev** [NVIDIA-NeMo/RL#4166]
## Goal

Match PrimeRL Example 5: full-parameter, text-only Nemotron 3.5 Super GRPO on Harbor SimpleQA, with 16 training GPUs and 16 inference GPUs. We need to offload saved activations to CPU while keeping model parameters and optimizer state on GPU.

| Setting | Completed PrimeRL Example 5 | NeMo attempt |
| --- | --- | --- |
| GPU placement | 16 training + 16 inference GPUs | Matched |
| Agent / task | Verifiers Bash + Harbor SimpleQA | Terminus2NemoGym + the same Harbor task split and grader |
| Parameter / optimizer CPU offload | Both off | Matched |
| Activation CPU offload | On, alongside checkpointing | No equivalent exposed in this AutoModel path; checkpointing on |
| Scheduling | Async; maximum policy age 1 | Async; maximum trajectory age 1; implementation differs |
| Batch / group size | 256 / 8 | Matched target; no dynamic replacement |
| Training parallelism | FSDP2 across 16 training GPUs: DP shard 16, DP replicate 1; TP / CP / EP = 1 / 1 / 1 | Matched |
| Precision | BF16 compute; FP32 stored parameters and optimization | FP32 stored parameters; initially FP32 moments, then BF16 retries |
| LR / betas / weight decay / grad norm cap | 1e-5 / 0.9, 0.999 / 0 / 1 | Matched |
| Context / output / maximum turns | 32K / 4K / 8 | Matched limits; actual conversations differ |
| Inference parallelism | 4 separate vLLM engines × TP4 = 16 GPUs; EP1, CUDA graphs, 0.85 memory fraction | Matched settings; different vLLM versions |
| Output-layer memory | Fused, chunked forward and backward | Full FP32 logits and training log-softmax |
| Validation | 16 before/after; temperature 0; interval 10 | Same settings; failed before final validation |
| Agent / verifier timeout | 600 / 120 seconds | Matched |
| Rollout concurrency | Adaptive 8–128, starting at 64 | Fixed maximum 128 |

FSDP2 shards parameters, gradients and optimizer state across the 16 training
ranks; each rank processes different trajectories. TP/CP/EP = 1 therefore still
uses distributed training, without tensor, context or expert parallelism.

## Description

The tested FSDP2/AutoModel path has activation checkpointing and model-state CPU offload, but no exposed activation-only CPU offload option. This prevents matching PrimeRL's memory configuration for long, multi-turn trajectories.

Checkpointing recomputes intermediate values during backward, but still retains inputs needed for recomputation. Activation offload can move eligible saved tensors to CPU until backward needs them, leaving more GPU memory for the next layer's temporary buffers. It trades host memory and transfer time for GPU capacity. [PyTorch documents this mechanism](https://docs.pytorch.org/docs/2.11/autograd.html#torch.autograd.graph.save_on_cpu).

These existing options address different objects or phases:

- `policy.dtensor_cfg.activation_checkpointing`: recomputation; already enabled.
- `policy.dtensor_cfg.cpu_offload`: FSDP model/gradient/optimizer-state offload.
- `policy.offload_optimizer_for_logprob`: optimizer-state offload around log-probability calculation.
- `policy.megatron_cfg.fine_grained_activation_offloading`: a Megatron option, outside this FSDP2 request.

## Environment and observed impact

- NeMo RL AutoModel setup from `6453bfd66bda1572654b679fb7a9194f3033b148`; AutoModel 0.7.0, PyTorch 2.11.0+cu130, Transformers 5.15.1.
- Eight GB200 nodes, four GPUs per node; 16 training ranks and 16 inference GPUs.
- BF16 compute, FP32 stored parameters and Adam moments, activation checkpointing enabled.
- Harbor SimpleQA through Terminus2NemoGym; 256 trajectories/update, 32K context, native dynamic batching with a 32,768-token microbatch budget; packing disabled.

The September 17 run completed **1/3 updates**, then failed in the second update's MoE forward activation (`relu2_deepep`), before the vocabulary projection:

| Recorded memory at failure | Value |
| --- | --- |
| Requested BF16 intermediate | 3.44 GiB |
| Free GPU memory | 2.83 GiB |
| PyTorch allocated | 172.34 GiB |
| Reserved but unused | 115.73 MiB |

PrimeRL 0.9.0 completed **3/3 updates plus final validation**, with activation offload and full checkpointing enabled, and parameter/optimizer offload disabled. Its maximum reported trainer-rank-0 PyTorch reserved memory was **157.4 GiB**.

PrimeRL also uses sequence packing and a tiled output layer; trajectories and kernels differ. We have not isolated activation offload's contribution or shown that it alone fixes this OOM. The failure motivates a missing capability; it is not evidence that dynamic batching itself is incorrect. The [retained findings](https://gitlab-master.nvidia.com/xiyu/benchling-buidathon-super-nemorl/-/blob/xinyu-dev/super35-nemorl/examples/lyris/super35/README.md#packing-and-dynamic-batching-checks--september-17) contain the run summary; failed-run configs and raw logs were removed during cleanup.

## Steps to reproduce the configuration gap

1. Inspect NeMo RL at the revision above, using its FSDP2/AutoModel worker (`policy.dtensor_cfg._v2: true`).
2. Keep checkpointing enabled and model/optimizer offload disabled:

   ```yaml
   policy:
     offload_optimizer_for_logprob: false
     dtensor_cfg:
       enabled: true
       _v2: true
       activation_checkpointing: true
       cpu_offload: false
       tensor_parallel_size: 1
       context_parallel_size: 1
       expert_parallel_size: 1
   ```

3. Inspect the [AutoModel setup](https://github.com/NVIDIA-NeMo/RL/blob/6453bfd66bda1572654b679fb7a9194f3033b148/nemo_rl/models/automodel/setup.py#L475-L485): it wires `CPUOffloadPolicy` and activation checkpointing, but has no corresponding saved-activation offload control.
4. Compare PrimeRL's completed run settings: `model.ac.mode="full"`, `model.ac_offloading.pin_memory=true`, `model.fsdp_cpu_offload=false`, `model.optim_cpu_offload=false`. NeMo's `cpu_offload=true` does not express the same configuration because it changes model-state residency.

These steps verify the missing configuration capability; the excerpt is not a complete OOM reproducer. The retained runnable NeMo example is the successful single-turn baseline.

## Expected behavior

- Expose a documented, independent activation CPU offload option for FSDP2/AutoModel, compatible with activation checkpointing and the native Nemotron model.
- Keep parameter and optimizer placement unchanged when enabling it.
- Validate loss/gradient agreement with offload disabled on identical inputs, then exercise at least three optimizer updates, including after Adam state is initialized.
- Report peak GPU/host memory and step-time tradeoffs; a particular memory reduction is not assumed.

## References

- [PrimeRL 0.9.0 activation offload implementation](https://github.com/PrimeIntellect-ai/prime-rl/blob/v0.9.0/src/prime_rl/utils/act_offloading.py)
- [Completed PrimeRL resolved trainer config](https://github.com/xinyu-dev/benchling-buidathon-super-primerl/blob/222b361ac85c708fbdddd27ba0011fbfc4026697/outputs/nemotron35/grpo-r3-harbor-smoke/configs/resolved/trainer.json) and [result summary](https://github.com/xinyu-dev/benchling-buidathon-super-primerl/blob/222b361ac85c708fbdddd27ba0011fbfc4026697/outputs/nemotron35/grpo-r3-harbor-smoke/summary.json) (repository access required).
- [Related harness-parity request #4165](https://github.com/NVIDIA-NeMo/RL/issues/4165).
